Renewable + PEM Parameterized Bidder

Xinhe-Chen opened the pull request:
## Fixes
Add the renewable+PEM parameterized bidder (bids the power at a constant price).

## Summary/Motivation:
1. Add a PerfectForecaster to predict the day-ahead and real-time wind power and electricity prices (if needed).
2. Add a PEMParameterizedd bidder to bid the renewable power at a constant prices.

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
3. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

andrewlee94 commented on idaes/apps/grid_integration/tests/test_perfectforecaster.py:
```diff
@@ -0,0 +1,121 @@
+    pyo_unittest.assertStructuredAlmostEqual(
+        first=result_forecast.tolist(), second=expected_forecast
+    )
+    return
```
Tests should not return anything - this causes a lot of warnings to be emitted in the test output which makes it hard to find real issues. There are 4 more instances of this below as well.

andrewlee94 commented on idaes/apps/grid_integration/tests/test_perfectforecaster.py:
```diff
@@ -0,0 +1,121 @@
+@pytest.mark.unit
+@pytest.mark.parametrize("value", [np.array([1,2,3,4,5]), [1,2,3,4,5]])
+def test_create_perfectforecaster_with_ndarray_and_list(value):
+    with pytest.raises(ValueError):
```
When checking for errors, you should also include the expected text message to ensure you got the right error (and not another of the same type).

andrewlee94 commented on idaes/apps/grid_integration/tests/test_perfectforecaster.py:
```diff
@@ -0,0 +1,121 @@
+    return PerfectForecaster(wind_df)
+
+@pytest.mark.unit
+def test_create_perfectforecaster(wind_df):
```
These are more like `component` tests than `unit` tests. Unit tests should generally be focused on individual methods/functions in the code to make sure all the elements work correctly in isolation, followed by a few test cases like these to make sure the end-to-end workflow is correct.

adowling2 replied:
@Xinhe-Chen, I think you already did this. Is so, just reply here with a short sentence confirming/summarizing what you did.

Xinhe-Chen replied:
I have addressed this.

andrewlee94 commented on idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:
```diff
@@ -0,0 +1,72 @@
+    TestingForecaster,
+    testing_model_data,
+)
+from pyomo.common import unittest as pyo_unittest
```
Group imports together, starting with native Python packages, then third-party packages, then Pyomo and finally IDAES.

andrewlee94 commented on idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:
```diff
@@ -0,0 +1,72 @@
+
+day_ahead_horizon = 24
+real_time_horizon = 4
+solver = pyo.SolverFactory("cbc")
```
You should probably add a check to ensure `cbc` is available in the test environment and to skip these tests if it is not.

andrewlee94 commented on idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:
```diff
@@ -0,0 +1,72 @@
+    return bidder_object
+
+
+# @pytest.mark.unit
```
Commented code.

Also, there is not a lot of testing for this class here - there are a number of methods that should be unit tested.

andrewlee94 reviewed (changes requested):
A number of requests regarding improvements to testing.

andrewlee94 commented:
Also, we will need documentation for the new classes and how to use them.

Xinhe-Chen commented:
> Also, we will need documentation for the new classes and how to use them.

Thank you for your comments! I am still working on improving the tests. The documentation will be added in this PR https://github.com/IDAES/idaes-pse/pull/1090.

Xinhe-Chen commented:
@radhakrishnatg Hi Radhakrishna, would you please review this PR? Thank you!

Xinhe-Chen commented:
@adowling2 this is ready for review.

adowling2 reviewed (commented):
Do the unit tests check all/any error messages thrown (e.g., input validation)?

andrewlee94 commented on idaes/apps/grid_integration/tests/test_perfectforecaster.py:
```diff
@@ -0,0 +1,137 @@
+@pytest.mark.unit
+@pytest.mark.parametrize("value", [np.array([1, 2, 3, 4, 5]), [1, 2, 3, 4, 5]])
+def test_create_perfectforecaster_with_ndarray_and_list(value):
+    with pytest.raises(
```
Tests should have some documentation so the next maintainer knows what they are doing. I try to make the name somewhat indicative, and put inline comments where necessary.

Remember that these are there to remind you when you have to work out why something broke in the future (or help someone else understand).

ksbeattie commented:
@Xinhe-Chen any news on this?

Xinhe-Chen commented:
> @Xinhe-Chen any news on this?

I will request a review from Prof. Dowling again.

adowling2 commented:
Tag other people too

dguittet commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,390 @@ def _record_bids(self, bids, date, hour, **kwargs):
+        Check the power of PEM should not exceed the power of renewables
+        """
+        if self.pem_mw >= self.renewable_mw:
+            raise ValueError(f"The power of PEM is greater than the renewabele power.")
```
Typo in renewable

dguittet commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,390 @@ def _record_bids(self, bids, date, hour, **kwargs):
+        for t_idx in range(self.day_ahead_horizon):
+            da_wind = forecast[t_idx] * self.renewable_mw
+            grid_wind = max(0, da_wind - self.pem_mw)
+            # gird wind are bidded at marginal cost = 0
```
typo in grid

dguittet commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,390 @@ def _record_bids(self, bids, date, hour, **kwargs):
+                    # When having indexerror, it must be the period that we are looking ahead. It is ok to set da_dispatch to 0
+                    da_dispatch = 0
+            # if we only participates in the RT market, then we do not consider the DA commitment
+            if self.real_time_bidding_only:
```
this could be just "else"

dguittet commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,390 @@ def _record_bids(self, bids, date, hour, **kwargs):
+                bids = [(0, 0), (0, 0)]
+            if avail_rt_wind > 0 and grid_wind == 0:
+                bids = [(0, 0), (avail_rt_wind, self.pem_marginal_cost)]
+            if avail_rt_wind > 0 and grid_wind > 0:
```
shouldn't this use a if-else structure? 
```
if avail_rt_wind == 0:
    bids = [(0, 0), (0, 0)]
else:
    if grid_wind == 0:
        bids = [(0, 0), (avail_rt_wind, self.pem_marginal_cost)]
    else:
        bids = [(0, 0), (grid_wind, 0), (avail_rt_wind, self.pem_marginal_cost)]
```

dguittet commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,390 @@ def _record_bids(self, bids, date, hour, **kwargs):
+            if avail_rt_wind > 0 and grid_wind > 0:
+                bids = [(0, 0), (grid_wind, 0), (avail_rt_wind, self.pem_marginal_cost)]
+            cost_curve = convert_marginal_costs_to_actual_costs(bids)
+            print(bids)
```
remove print

dguittet commented on idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:
```diff
@@ -0,0 +1,198 @@
+                (expected_DA_cf[t] * pmax, pem_marginal_cost),
+            ]
+        expect_cost_curve = convert_marginal_costs_to_actual_costs(expect_bids_curve)
+        expected_bids[t] = {
```
This test is re-implementing the code/logic of the bidder. While this is possible because the logic is straightforward and simple, it seems to me like the test should use actual numbers for the expected bid curves so that the test pass/fail is not implementation-specific

adowling2 replied:
Was this comment addressed? If so, reply here.

adowling2 replied:
@Xinhe-Chen Please reply here if you addressed this.

Xinhe-Chen replied:
Yes this has been addressed by me.

dguittet commented on idaes/apps/grid_integration/tests/test_PEM_Parameterized_bidder.py:
```diff
@@ -0,0 +1,198 @@
+        else:
+            expect_bids_curve = [(0, 0)]
+        expect_cost_curve = convert_marginal_costs_to_actual_costs(expect_bids_curve)
+        expected_bids[t] = {
```
Same comment as above

Xinhe-Chen replied:
Hi Darice, thank you for spending time reviewing the codes! I have addressed all your comments today.

Xinhe-Chen commented:
@adowling2 This is ready for review.

Xinhe-Chen commented:
@ksbeattie Hi Keith, we hope to have this PR reviewed and merged next week.

adowling2 commented on docs/reference_guides/apps/grid_integration/Bidder.rst:
```diff
@@ -13,6 +13,155 @@ uncertain price scenario has a corresponding power output. As shown in the figur
 each of these uncertain price and power output pairs formulates a segment in the
 bidding curves.
 
+Here we present a stochastic bidding model for a renewable integrated energy system (Wind generator + PEM).
```
wind should be lowercase, make sure you define PEM on first use

Xinhe-Chen replied:
This is addressed.

adowling2 commented on docs/reference_guides/apps/grid_integration/Bidder.rst:
```diff
@@ -13,6 +13,155 @@ uncertain price scenario has a corresponding power output. As shown in the figur
+in order to avoid underbidding. Equation (2) states that the RT offering power is the same as the 
+IES power output to the grid. In the bidding mode, the market rules require the offering power is 
+non-decreasing (convex) with its marginal cost in an energy bid. This rule is represented by equation (3).
+Equation (4) to equation (9) are process model constraints. Equation (10) calculates the operation costs for IES
```
Equations (4) to (9)

Xinhe-Chen replied:
Done.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+        This function records the bids and the details in the underlying bidding model.
+
+        Arguments:
+            bids: the obtained bids for this date.
```
What is the specific data structure for a bid? Is this a list? Please provide a little more information in the comments

andrewlee94 replied:
Type hinting would be a good idea. E.g. the method signature could be:

```
def record_bids(self, bids: list, model, date, hour, market):
```

Xinhe-Chen replied:
I added variable types in code comments.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+
+            model: bidding model
+
+            date: the date we bid into
```
Is this a string or something else?

andrewlee94 replied:
Typing hinting will help again. Also, remove the line breaks between each argument (otherwise the docs formatter might get confused).

Xinhe-Chen replied:
Done

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+
+            date: the date we bid into
+
+            hour: the hour we bid into
```
Is this an integer or something else?

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+            None
+
+        """
+        df_list = []
```
Please add a few line comments throughout this code to explain what is happening.

adowling2 replied:
@Xinhe-Chen, did you address this comment? If so, please reply here.

Xinhe-Chen replied:
Done

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+                for k, v in kwargs.items():
+                    result_dict[k] = v
+
+                pair_cnt = len(bids[t][gen]["p_cost"])
```
What does `pair_cnt` stand for? The number of bid elements? Perhaps change to `num_bid_pairs` as the variable name?

Xinhe-Chen replied:
Done

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+                    result_dict[f"Power {idx} [MW]"] = power
+                    result_dict[f"Cost {idx} [$]"] = cost
+
+                # place holder, in case different len of bids
```
Does this code pad the bid to ensure it has the correct length? If so, please say this more clearly in the comments.

Xinhe-Chen replied:
I explained this at Line 1329. The bid length is correct.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+    ):
+        """
+        Arguments:
+            renewable_mw: maximum renewable energy system capacity
```
Is this doc string missing arguments (such as bidding_model_object, day_aheady_horizon, ...)?

Xinhe-Chen replied:
Those are inherited from the 'ParameterizedBidder'. I have added the doc strings for them now.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+    ):
+        """
+        Arguments:
+            renewable_mw: maximum renewable energy system capacity
```
Is this a float? If so, add (float) or , float to the end of the doc string.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+        Arguments:
+            renewable_mw: maximum renewable energy system capacity
+
+            pem_marginal_cost: the cost/MW above which all available wind energy will be sold to grid;
```
Is this a float?

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,395 @@ def _record_bids(self, bids, date, hour, **kwargs):
+            pem_marginal_cost: the cost/MW above which all available wind energy will be sold to grid;
+                below which, make hydrogen and sell remainder of wind to grid
+
+            pem_mw: maximum PEM capacity limits how much energy is bid at the `pem_marginal_cost`
```
Is this a float?

adowling2 commented on idaes/apps/grid_integration/forecaster.py:
```diff
@@ -681,3 +681,57 @@ def fetch_day_ahead_stats_from_prescient(self, uc_date, uc_hour, day_ahead_resul
+         - {bus}-DALMP
+         - {bus}-RTLMP
+         - {bus}-DACF and {bus}-RTCF for renewable plants
+        """
```
Perhaps add a sentence explaining the perfect forecast has no error and "sees into the future".

adowling2 replied:
@Xinhe-Chen, did you do this? If so, please reply here.

Xinhe-Chen replied:
Done

adowling2 reviewed (changes requested):
Please see the comments below. These are all asking for minor changes, often just more descriptive comments.

adowling2 reviewed (commented):
@Xinhe-Chen, please reply to the open comments to confirm these were addressed.

Xinhe-Chen commented:
> @Xinhe-Chen, please reply to the open comments to confirm these were addressed.

Hi Prof. Dowling, I addressed all comments from you last Friday. 

I am going to to a final check later today to ensure no other issues are remaining later today.

adowling2 commented:
Excellent. Tag me here when it is ready to review.

Xinhe-Chen commented:
> Excellent. Tag me here when it is ready to review.

@adowling2 This is ready for review.

adowling2 reviewed (approved):
Ready to merge.

andrewlee94 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -1291,3 +1291,413 @@ def _record_bids(self, bids, date, hour, **kwargs):
+            None
+        """
+
+        print("")
```
Using a logger is generally preferred to doing `print` statements as this gives the end-user more flexibility in how (and what) things are displayed.

Also, for a new line use `print("\n"}`.

andrewlee94 replied:
https://github.com/IDAES/idaes-pse/blob/main/.github/CODEOWNERS

Xinhe-Chen replied:
I saw @adowling2 is on the CODEOWNERS roster for the grid_integration codes. I am a PhD student and will graduate in 2 years but I am happy to be responsible for maintaining these codes until I graduate.

Xinhe-Chen replied:
> Using a logger is generally preferred to doing `print` statements as this gives the end-user more flexibility in how (and what) things are displayed.
> 
> Also, for a new line use `print("\n"}`.

Done.

andrewlee94 replied:
I see that you changed the new line, but what about using a logger?

Xinhe-Chen replied:
I am not very familiar with the logger, so I need to look into that.

Xinhe-Chen replied:
Hi Andrew, I used logger to replace print and it works. Thank you for letting me know this.

andrewlee94 reviewed (commented):
@Xinhe-Chen I have one final request before we merge this; someone should be added to the top level CODEOWNERS file as the maintainer for this code (i.e. the person Keith or Ludovico will contact if tests start failing). We require all code in `apps` to have an assigned owner for maintenance purposes.

Xinhe-Chen commented:
@andrewlee94 Hi Andrew, I think I have already addressed your comments. Shall we get this PR merged?

andrewlee94 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -755,8 +759,8 @@ def write_results(self, path):
 
-        print("")
-        print("Saving bidding results to disk...")
+        _logger.info("\n")
```
You do not need the new line with the logger. Also, in the next line, I would use only a single period (`.`) instead of an elipsis (`...`).

andrewlee94 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -11,6 +11,7 @@
 # for full copyright and license information.
 #################################################################################
 import os
+import logging
```
Rather than the base logging package, could you use `import idaes.logger as idaeslog`

`_logger = idaeslog.getLogger(__name__)`

Also note that you do not need to set the logger level here, that is done by the user to control the output.

andrewlee94 reviewed (changes requested):
Sorry for the delay, I was at a conference or out on leave the past three weeks. A couple of minor requests for changes to how you are using the logger.

Xinhe-Chen commented:
> Sorry for the delay, I was at a conference or out on leave the past three weeks. A couple of minor requests for changes to how you are using the logger.

Thank you! I have addressed your new comments.
